pkg/controller/pxc/controller.go

```diff
@@ -1109,3 +1102,40 @@ func createOrUpdateConfigmap(cl client.Client, configMap *corev1.ConfigMap) erro
 
 	return nil
 }
+
+func isServiceNeedsToUpdate(cl client.Client, old, new *corev1.Service) bool {
+	if !isServicesSpecEqual(old.Spec, new.Spec) || !reflect.DeepEqual(old.Annotations, new.Annotations) ||
+		!reflect.DeepEqual(old.Labels, new.Labels) || !reflect.DeepEqual(new.Spec.Selector, old.Spec.Selector) {
+
+		return true
+	}
+	return false
+}
+
+func isServicesSpecEqual(old, new corev1.ServiceSpec) bool {
+	if old.Type != new.Type || !isServicesPortsEqual(old.Ports, new.Ports) ||
+		old.ExternalTrafficPolicy != new.ExternalTrafficPolicy ||
+		!reflect.DeepEqual(old.LoadBalancerSourceRanges, new.LoadBalancerSourceRanges) {
+
+		return false
+	}
+	return true
+}
+
+func isServicesPortsEqual(old, new []corev1.ServicePort) bool {
+	if len(old) != len(new) {
+		return false
+	}
+
+	for _, oldSpec := range old {
+		for _, newSpec := range new {
+			if oldSpec.Name == newSpec.Name {
+				if oldSpec.Port != newSpec.Port || oldSpec.Protocol != newSpec.Protocol ||
+					oldSpec.TargetPort.IntVal != newSpec.TargetPort.IntVal {
+					return false
+				}
+			}
+		}
+	}
+	return true
```

Comment:
It's easier to sort it by name and compare index by index.
Wnat if we have:
[name1,name2] and [name2,name3] ? 
lens are equal but names are not.
your algorithm will return true. But as I understand it should return false

Reply:
This is actually a good example of func we can write test for

Reply:
nice catch
5f4fc8dd